src/data.cc

```diff
@@ -111,50 +111,120 @@ DMLC_REGISTER_PARAMETER(CSVParserParam);
 }
 
 template<>
-Parser<uint64_t> *
-Parser<uint64_t>::Create(const char *uri_,
-                         unsigned part_index,
-                         unsigned num_parts,
-                         const char *type) {
-  return data::CreateParser_<uint64_t>(uri_, part_index, num_parts, type);
+RowBlockIter<uint64_t, int> *
+RowBlockIter<uint64_t, int>::Create(const char *uri,
+                                    unsigned part_index,
+                                    unsigned num_parts,
+                                    const char *type) {
+  return data::CreateIter_<uint64_t, int>(uri, part_index, num_parts, type);
 }
 
-// registry
-DMLC_REGISTRY_ENABLE(ParserFactoryReg<uint32_t>);
-DMLC_REGISTRY_ENABLE(ParserFactoryReg<uint64_t>);
-DMLC_REGISTER_DATA_PARSER(uint32_t, libsvm, data::CreateLibSVMParser<uint32_t>);
-DMLC_REGISTER_DATA_PARSER(uint64_t, libsvm, data::CreateLibSVMParser<uint64_t>);
+template<>
+Parser<uint32_t, real_t> *
+Parser<uint32_t, real_t>::Create(const char *uri_,
+                                 unsigned part_index,
+                                 unsigned num_parts,
+                                 const char *type) {
+  return data::CreateParser_<uint32_t, real_t>(uri_, part_index, num_parts, type);
+}
+
+template<>
+Parser<uint64_t, real_t> *
+Parser<uint64_t, real_t>::Create(const char *uri_,
+                                 unsigned part_index,
+                                 unsigned num_parts,
+                                 const char *type) {
+  return data::CreateParser_<uint64_t, real_t>(uri_, part_index, num_parts, type);
+}
+
+template<>
+Parser<uint32_t, int> *
+Parser<uint32_t, int>::Create(const char *uri_,
+                              unsigned part_index,
+                              unsigned num_parts,
+                              const char *type) {
+  return data::CreateParser_<uint32_t, int>(uri_, part_index, num_parts, type);
+}
 
-DMLC_REGISTER_DATA_PARSER(uint32_t, libfm, data::CreateLibFMParser<uint32_t>);
-DMLC_REGISTER_DATA_PARSER(uint64_t, libfm, data::CreateLibFMParser<uint64_t>);
+template<>
+Parser<uint64_t, int> *
+Parser<uint64_t, int>::Create(const char *uri_,
+                              unsigned part_index,
+                              unsigned num_parts,
+                              const char *type) {
+  return data::CreateParser_<uint64_t, int>(uri_, part_index, num_parts, type);
+}
 
-DMLC_REGISTER_DATA_PARSER(uint32_t, csv, data::CreateCSVParser<uint32_t>);
+// registry
+typedef ParserFactoryReg<uint32_t, real_t> Reg32flt;
+typedef ParserFactoryReg<uint32_t, int> Reg32int;
+typedef ParserFactoryReg<uint64_t, real_t> Reg64flt;
+typedef ParserFactoryReg<uint64_t, int> Reg64int;
+DMLC_REGISTRY_ENABLE(Reg32flt);
+DMLC_REGISTRY_ENABLE(Reg32int);
+DMLC_REGISTRY_ENABLE(Reg64flt);
+DMLC_REGISTRY_ENABLE(Reg64int);
+
+static DMLC_ATTRIBUTE_UNUSED ParserFactoryReg<uint32_t, real_t> &
+  __make_ParserFactoryReg_uint32_t_real_t_libsvm__ =
```

Comment:
Please check if this method helps us use the existing macros and avoid listing out the definitions.

https://stackoverflow.com/a/19841470/2382819

Reply:
Done